handle missingness correctly for arrays in user interface

Example that currently fails:

```
>>> mt = hl.import_vcf('sample.vcf')
>>> mt.annotate_entries(PL = hl.cond(mt.DP > 20, mt.PL, hl.map(lambda x: x, mt.PL)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/cotton/hail/python/hail/matrixtable.py", line 975, in annotate_entries
    return self._select_entries(caller, s=self.entry.annotate(**e))
  File "/home/cotton/hail/python/hail/typecheck/check.py", line 547, in wrapper
    return f(*args_, **kwargs_)
  File "/home/cotton/hail/python/hail/matrixtable.py", line 2887, in _select_entries
    return cleanup(MatrixTable(base._jvds.selectEntries(str(s._ir))))
  File "/home/cotton/opt/spark-2.2.0-bin-hadoop2.7/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
  File "/home/cotton/hail/python/hail/utils/java.py", line 212, in deco
    'Error summary: %s' % (deepest, full, hail.__version__, deepest)) from None
hail.utils.java.FatalError: AssertionError: assertion failed: mismatch:
  array<int32>
  array<int32>
  ApplyComparisonOp(GT(int32,int32),GetField(ArrayRef(GetField(Ref(va,struct{locus: locus<GRCh37>, alleles: array<str>, rsid: str, qual: float64, filters: set<str>, info: struct{NEGATIVE_TRAIN_SITE: bool, HWP: float64, AC: array<int32>, culprit: str, MQ0: int32, ReadPosRankSum: float64, AN: int32, InbreedingCoeff: float64, AF: array<float64>, GQ_STDDEV: float64, FS: float64, DP: int32, GQ_MEAN: float64, POSITIVE_TRAIN_SITE: bool, VQSLOD: float64, ClippingRankSum: float64, BaseQRankSum: float64, MLEAF: array<float64>, MLEAC: array<int32>, MQ: float64, QD: float64, END: int32, DB: bool, HaplotypeScore: float64, MQRankSum: float64, CCC: int32, NCC: int32, DS: bool}, `the entries! [877f12a8827e18f61222c6c8c5fb04a8]`: array<struct{GT: call, AD: array<int32>, DP: int32, GQ: int32, PL: array<int32>}>}),the entries! [877f12a8827e18f61222c6c8c5fb04a8]),Ref(i,int32)),DP),I32(20))

Java stack trace:
java.lang.AssertionError: assertion failed: mismatch:
  array<int32>
  array<int32>
  ApplyComparisonOp(GT(int32,int32),GetField(ArrayRef(GetField(Ref(va,struct{locus: locus<GRCh37>, alleles: array<str>, rsid: str, qual: float64, filters: set<str>, info: struct{NEGATIVE_TRAIN_SITE: bool, HWP: float64, AC: array<int32>, culprit: str, MQ0: int32, ReadPosRankSum: float64, AN: int32, InbreedingCoeff: float64, AF: array<float64>, GQ_STDDEV: float64, FS: float64, DP: int32, GQ_MEAN: float64, POSITIVE_TRAIN_SITE: bool, VQSLOD: float64, ClippingRankSum: float64, BaseQRankSum: float64, MLEAF: array<float64>, MLEAC: array<int32>, MQ: float64, QD: float64, END: int32, DB: bool, HaplotypeScore: float64, MQRankSum: float64, CCC: int32, NCC: int32, DS: bool}, `the entries! [877f12a8827e18f61222c6c8c5fb04a8]`: array<struct{GT: call, AD: array<int32>, DP: int32, GQ: int32, PL: array<int32>}>}),the entries! [877f12a8827e18f61222c6c8c5fb04a8]),Ref(i,int32)),DP),I32(20))
	at scala.Predef$.assert(Predef.scala:170)
	at is.hail.expr.ir.Infer$.apply(Infer.scala:10)
	at is.hail.expr.ir.InferIR$class.typ(IR.scala:58)
	at is.hail.expr.ir.If.typ(IR.scala:77)
	at is.hail.expr.ir.Infer$$anonfun$apply$5.apply(Infer.scala:63)
	at is.hail.expr.ir.Infer$$anonfun$apply$5.apply(Infer.scala:59)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at is.hail.expr.ir.Infer$.apply(Infer.scala:59)
	at is.hail.expr.ir.InferIR$class.typ(IR.scala:58)
	at is.hail.expr.ir.InsertFields.typ(IR.scala:154)
	at is.hail.expr.ir.InsertFields.typ(IR.scala:153)
	at is.hail.expr.ir.Infer$.apply(Infer.scala:39)
	at is.hail.expr.ir.InferIR$class.typ(IR.scala:58)
	at is.hail.expr.ir.ArrayMap.typ(IR.scala:101)
	at is.hail.expr.ir.ArrayMap.typ(IR.scala:100)
	at is.hail.expr.ir.Infer$$anonfun$apply$5.apply(Infer.scala:63)
	at is.hail.expr.ir.Infer$$anonfun$apply$5.apply(Infer.scala:59)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at is.hail.expr.ir.Infer$.apply(Infer.scala:59)
	at is.hail.expr.ir.InferIR$class.typ(IR.scala:58)
	at is.hail.expr.ir.InsertFields.typ(IR.scala:154)
	at is.hail.expr.ir.MatrixMapEntries.<init>(MatrixIR.scala:1075)
	at is.hail.variant.MatrixTable.selectEntries(MatrixTable.scala:641)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)

Hail version: devel-0eb009391fd5
Error summary: AssertionError: assertion failed: mismatch:
  array<int32>
  array<int32>
  ApplyComparisonOp(GT(int32,int32),GetField(ArrayRef(GetField(Ref(va,struct{locus: locus<GRCh37>, alleles: array<str>, rsid: str, qual: float64, filters: set<str>, info: struct{NEGATIVE_TRAIN_SITE: bool, HWP: float64, AC: array<int32>, culprit: str, MQ0: int32, ReadPosRankSum: float64, AN: int32, InbreedingCoeff: float64, AF: array<float64>, GQ_STDDEV: float64, FS: float64, DP: int32, GQ_MEAN: float64, POSITIVE_TRAIN_SITE: bool, VQSLOD: float64, ClippingRankSum: float64, BaseQRankSum: float64, MLEAF: array<float64>, MLEAC: array<int32>, MQ: float64, QD: float64, END: int32, DB: bool, HaplotypeScore: float64, MQRankSum: float64, CCC: int32, NCC: int32, DS: bool}, `the entries! [877f12a8827e18f61222c6c8c5fb04a8]`: array<struct{GT: call, AD: array<int32>, DP: int32, GQ: int32, PL: array<int32>}>}),the entries! [877f12a8827e18f61222c6c8c5fb04a8]),Ref(i,int32)),DP),I32(20))
```

--- a/src/main/scala/is/hail/expr/ir/Compile.scala
+++ b/src/main/scala/is/hail/expr/ir/Compile.scala
@@ -41,6 +41,7 @@ object Compile {
 
     var ir = body
     ir = Optimize(ir, noisy = false)
+    TypeCheck(ir, Env.empty[Type].bind(args.map { case (name, t, _) => name -> t}: _*), None)
 
     val env = args
       .zipWithIndex

--- a/src/main/scala/is/hail/expr/ir/Infer.scala
+++ b/src/main/scala/is/hail/expr/ir/Infer.scala
@@ -7,6 +7,7 @@ object Infer {
     ir match {
       case If(cond, cnsq, altr) =>
         assert(cond.typ.isOfType(TBoolean()))
+        assert(cnsq.typ.isOfType(altr.typ))
         if (cnsq.typ != altr.typ)
           cnsq.typ.deepOptional()
         else

--- a/src/main/scala/is/hail/expr/ir/Infer.scala
+++ b/src/main/scala/is/hail/expr/ir/Infer.scala
@@ -7,9 +7,10 @@ object Infer {
     ir match {
       case If(cond, cnsq, altr) =>
         assert(cond.typ.isOfType(TBoolean()))
-        assert(cnsq.typ == altr.typ, s"mismatch:\n  ${ cnsq.typ.parsableString() }\n  ${ altr.typ.parsableString() }\n  $cond")
-        cnsq.typ
-
+        if (cnsq.typ != altr.typ)
+          cnsq.typ.deepOptional()
+        else
+          cnsq.typ
       case Let(name, value, body) =>
         body.typ
       case ApplyBinaryPrimOp(op, l, r) =>

--- a/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala
+++ b/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala
@@ -1006,8 +1006,9 @@ object PruneDeadFields {
         case ta: TArray =>
           val ra = rType.asInstanceOf[TArray]
           val uid = genUID()
-          val ref = Ref(uid, ta.elementType)
+          val ref = Ref(uid, -ta.elementType)
           ArrayMap(ir, uid, upcast(ref, ra.elementType))
+        case t => ir
       }
     }
   }

--- a/src/main/scala/is/hail/expr/ir/Simplify.scala
+++ b/src/main/scala/is/hail/expr/ir/Simplify.scala
@@ -104,8 +104,8 @@ object Simplify {
 
       case Let(n1, v, Ref(n2, _)) if n1 == n2 => v
 
-      case If(True(), x, _) => x
-      case If(False(), _, x) => x
+      case x@If(True(), cnsq, _) if x.typ == cnsq.typ => cnsq
+      case x@If(False(), _, altr) if x.typ == altr.typ => altr
 
       case If(c, cnsq, altr) if cnsq == altr =>
         if (cnsq.typ.required)

--- a/src/main/scala/is/hail/expr/ir/TypeCheck.scala
+++ b/src/main/scala/is/hail/expr/ir/TypeCheck.scala
@@ -5,16 +5,25 @@ import is.hail.utils._
 
 object TypeCheck {
   def apply(ir: IR, aggEnv: Option[Env[Type]] = None) {
+    apply(ir, new Env[Type](), aggEnv)
     try {
-      apply(ir, new Env[Type](), aggEnv)
     } catch {
-      case e: Throwable => fatal(s"Error while typechecking IR:\n${Pretty(ir)}", e)
+      case e: Throwable => fatal(s"Error while typechecking IR:\n${ Pretty(ir) }", e)
     }
   }
 
-  def apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]) {
+  def apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]): Unit = {
+    try {
+      _apply(ir, env, aggEnv)
+    } catch {
+      case e: Throwable => fatal(s"Error while typechecking IR:\n${ Pretty(ir) }", e)
+    }
+
+  }
+
+  private def _apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]) {
     def check(ir: IR, env: Env[Type] = env, aggEnv: Option[Env[Type]] = aggEnv) {
-      apply(ir, env, aggEnv)
+      _apply(ir, env, aggEnv)
     }
 
     ir match {
@@ -178,7 +187,7 @@ object TypeCheck {
         }: _*))
       case x@SelectFields(old, fields) =>
         check(old)
-        assert{
+        assert {
           val oldfields = coerce[TStruct](old.typ).fieldNames.toSet
           fields.forall { id => oldfields.contains(id) }
         }

--- a/src/main/scala/is/hail/expr/ir/TypeCheck.scala
+++ b/src/main/scala/is/hail/expr/ir/TypeCheck.scala
@@ -6,10 +6,6 @@ import is.hail.utils._
 object TypeCheck {
   def apply(ir: IR, aggEnv: Option[Env[Type]] = None) {
     apply(ir, new Env[Type](), aggEnv)
-    try {
-    } catch {
-      case e: Throwable => fatal(s"Error while typechecking IR:\n${ Pretty(ir) }", e)
-    }
   }
 
   def apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]): Unit = {
@@ -18,7 +14,6 @@ object TypeCheck {
     } catch {
       case e: Throwable => fatal(s"Error while typechecking IR:\n${ Pretty(ir) }", e)
     }
-
   }
 
   private def _apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]) {

--- a/src/main/scala/is/hail/expr/types/Type.scala
+++ b/src/main/scala/is/hail/expr/types/Type.scala
@@ -295,6 +295,7 @@ abstract class Type extends BaseType with Serializable {
       case t2: TArray => t.isInstanceOf[TArray] && t.asInstanceOf[TArray].elementType.isOfType(t2.elementType)
       case t2: TSet => t.isInstanceOf[TSet] && t.asInstanceOf[TSet].elementType.isOfType(t2.elementType)
       case t2: TDict => t.isInstanceOf[TDict] && t.asInstanceOf[TDict].keyType.isOfType(t2.keyType) && t.asInstanceOf[TDict].valueType.isOfType(t2.valueType)
+      case TVoid => t == TVoid
     }
   }
 

--- a/src/test/scala/is/hail/expr/ir/IRSuite.scala
+++ b/src/test/scala/is/hail/expr/ir/IRSuite.scala
@@ -141,10 +141,10 @@ class IRSuite extends SparkSuite {
     val value = Row(FastIndexedSeq(1, 2, 3))
     assertEvalsTo(
       If(
-        True(),
-        In(0, t),
+        In(0, TBoolean()),
+        In(1, t),
         MakeStruct(Seq("foo" -> ArrayRange(I32(0), I32(1), I32(1))))),
-      FastIndexedSeq((value, t)),
+      FastIndexedSeq((true, TBoolean()), (value, t)),
       value
     )
   }

--- a/src/test/scala/is/hail/expr/ir/IRSuite.scala
+++ b/src/test/scala/is/hail/expr/ir/IRSuite.scala
@@ -137,13 +137,13 @@ class IRSuite extends SparkSuite {
   }
 
   @Test def testIfWithDifferentRequiredness() {
-    val t = TStruct("foo" -> TArray(TInt32Required, required = true))
-    val value = Row(FastIndexedSeq(1, 2, 3))
+    val t = TStruct(true, "foo" -> TStruct("bar" -> TArray(TInt32Required, required = true)))
+    val value = Row(Row(FastIndexedSeq(1, 2, 3)))
     assertEvalsTo(
       If(
         In(0, TBoolean()),
         In(1, t),
-        MakeStruct(Seq("foo" -> ArrayRange(I32(0), I32(1), I32(1))))),
+        MakeStruct(Seq("foo" -> MakeStruct(Seq("bar" -> ArrayRange(I32(0), I32(1), I32(1))))))),
       FastIndexedSeq((true, TBoolean()), (value, t)),
       value
     )

--- a/src/test/scala/is/hail/expr/ir/IRSuite.scala
+++ b/src/test/scala/is/hail/expr/ir/IRSuite.scala
@@ -136,6 +136,19 @@ class IRSuite extends SparkSuite {
     assertEvalsTo(If(True(), NA(TInt32()), I32(7)), null)
   }
 
+  @Test def testIfWithDifferentRequiredness() {
+    val t = TStruct("foo" -> TArray(TInt32Required, required = true))
+    val value = Row(FastIndexedSeq(1, 2, 3))
+    assertEvalsTo(
+      If(
+        True(),
+        In(0, t),
+        MakeStruct(Seq("foo" -> ArrayRange(I32(0), I32(1), I32(1))))),
+      FastIndexedSeq((value, t)),
+      value
+    )
+  }
+
   @Test def testLet() {
     assertEvalsTo(Let("v", I32(5), Ref("v", TInt32())), 5)
     assertEvalsTo(Let("v", NA(TInt32()), Ref("v", TInt32())), null)

